False invalid-overridden-method positive with nested property

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?:
    pip install pylint astroid --pre -U
-->

Originally reported in https://github.com/PyCQA/pylint/issues/3150#issuecomment-538673964 which was closed as the initially reported issue was fixed, but the one I added still remains.

### Steps to reproduce

```python
$ cat t.py
def my_property(func):
    @property
    def _wrapper(self):
        pass

    return _wrapper


class A:
    @property
    def foo():
        return True


class B(A):
    @my_property
    def foo():
        return False
```

<!--
Given a file `a.py`:
```python
# (Disable message unrelated to the bug)
# pylint: disable=missing-docstring,...

```

Given multiple files:
```
__init__.py
a.py
module\
    __init__.py
    b.py
    c.py
```
-->

### Current behavior

<!--
Result of `pylint a.py`:
```

```
-->
```shellsession
$ pylint ./t.py  | grep overridden
t.py:17:4: W0236: Method 'foo' was expected to be 'property', found it instead as 'method' (invalid-overridden-method)
```

### Expected behavior

No invalid-overridden-method

### pylint --version output

Result of `pylint --version` output:

```
pylint 2.7.4
astroid 2.5.3
Python 3.8.0 (default, Mar 30 2021, 18:43:10) 
[GCC 9.3.0]
```

<!--
Additional dependencies:
```
pandas==0.23.2
marshmallow==3.10.0
...
```
-->

--- a/ChangeLog
+++ b/ChangeLog
@@ -53,6 +53,10 @@ Release date: TBA
 
   Closes #4698
 
+* Fix ``invalid-overridden-method`` with nested property
+
+  Closes #4368
+
 
 What's New in Pylint 2.9.3?
 ===========================

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -71,6 +71,7 @@ from typing import (
 
 import _string
 import astroid
+import astroid.objects
 
 from pylint.constants import BUILTINS
 
@@ -794,6 +795,22 @@ def _is_property_decorator(decorator: astroid.Name) -> bool:
             for ancestor in inferred.ancestors():
                 if ancestor.name == "property" and ancestor.root().name == BUILTINS:
                     return True
+        elif isinstance(inferred, astroid.FunctionDef):
+            # If decorator is function, check if it has exactly one return
+            # and the return is itself a function decorated with property
+            returns: List[astroid.Return] = list(
+                inferred._get_return_nodes_skip_functions()
+            )
+            if len(returns) == 1 and isinstance(
+                returns[0].value, (astroid.Name, astroid.Attribute)
+            ):
+                inferred = safe_infer(returns[0].value)
+                if (
+                    inferred
+                    and isinstance(inferred, astroid.objects.Property)
+                    and isinstance(inferred.function, astroid.FunctionDef)
+                ):
+                    return decorated_with_property(inferred.function)
     return False
 
 

--- a/tests/functional/i/invalid/invalid_overridden_method.py
+++ b/tests/functional/i/invalid/invalid_overridden_method.py
@@ -1,4 +1,4 @@
-# pylint: disable=missing-docstring, too-few-public-methods
+# pylint: disable=missing-docstring,too-few-public-methods,disallowed-name,invalid-name,unused-argument
 import abc
 
 
@@ -76,3 +76,52 @@ class AbstractProperty:
     @abc.abstractmethod
     def prop(self):
         return
+
+
+# https://github.com/PyCQA/pylint/issues/4368
+# Decrator functions with a nested property decorator should still be
+# inferred as property.
+
+def my_property(func):
+    @property
+    def _wrapper(self):
+        pass
+    return _wrapper
+
+def not_a_property(func):
+    def _wrapper(self):
+        pass
+    return _wrapper
+
+def multiple_returns(func):
+    def _wrapper(self):
+        pass
+    if foobar:  # pylint: disable=undefined-variable
+        return False
+    return _wrapper
+
+class A:
+    @property
+    def foo(self):
+        return True
+
+    @property
+    def bar(self):
+        return True
+
+    @property
+    def bar2(self):
+        return True
+
+class B(A):
+    @my_property
+    def foo(self):
+        return False
+
+    @not_a_property
+    def bar(self):  # [invalid-overridden-method]
+        return False
+
+    @multiple_returns
+    def bar2(self):  # [invalid-overridden-method]
+        return False

--- a/tests/functional/i/invalid/invalid_overridden_method.txt
+++ b/tests/functional/i/invalid/invalid_overridden_method.txt
@@ -2,3 +2,5 @@ invalid-overridden-method:38:4:InvalidDerived.prop:Method 'prop' was expected to
 invalid-overridden-method:41:4:InvalidDerived.async_method:Method 'async_method' was expected to be 'async', found it instead as 'non-async'
 invalid-overridden-method:45:4:InvalidDerived.method_a:Method 'method_a' was expected to be 'method', found it instead as 'property'
 invalid-overridden-method:48:4:InvalidDerived.method_b:Method 'method_b' was expected to be 'non-async', found it instead as 'async'
+invalid-overridden-method:122:4:B.bar:Method 'bar' was expected to be 'property', found it instead as 'method'
+invalid-overridden-method:126:4:B.bar2:Method 'bar2' was expected to be 'property', found it instead as 'method'

